README.md
# Tmux Sessionizer

[![semantic-release: angular](https://img.shields.io/badge/semantic--release-angular-e10079?logo=semantic-release)](https://github.com/semantic-release/semantic-release)
![license](https://img.shields.io/github/license/augustinesaidimu/sessionizer)
![tag](https://img.shields.io/github/v/tag/augustinesaidimu/sessionizer?sort=semver)
![release](https://img.shields.io/github/workflow/status/augustinesaidimu/sessionizer/Release)

## Origins

The idea originally belongs to [ThePrimeagen](https://github.com/ThePrimeagen).
Read the original [here](https://github.com/ThePrimeagen/.dotfiles/blob/5cd09f06d6683b91c26822a73b40e3d7fb9af57a/bin/.local/bin/tmux-sessionizer) and see how it's used [here](https://github.com/ThePrimeagen/.dotfiles/blob/5cd09f06d6683b91c26822a73b40e3d7fb9af57a/tmux/.tmux.conf#L25)

## Algorithm:

**Pseudo Code** in JavaScript:

```javascript
function configureSession ({ name, path }){
    if (existsInPath(path, ".project")) {
        execute(path, ".project");
    }

    if (existsInPath(path, ".env")) {
        getEnv(path).forEach((v) =>
            tmux.setEnvironmentVar({
                variable: v,
                session: name,
            })
        );
    }
};

function sessionize(){
    const list = readOrGenerate();
    const { name, path } = getSelectionFrom(list);

    if (!tmux.hasSession(name)) {
        tmux.createSession(name);
        configureSession({ name, path });
    }

    switchToSession(name);
};
```

Put simply, the script allows the user to select a path from a list.
Using the selected path, it switches to an existing tmux session, or
creates and configures a session before switching to it. The are two ways
provided of populating the target list. Typing into a text file the list
of paths, or extending the tool by providing the implementation of
a generator. Configuring the session involves invoking an executible named
`.project` and setting session-wide environment variables as defined in
a `.env` file within the target directory.

## Usage

#### Installation

```
sh <(curl -fsSL https://tinyurl.com/4t2cbsh8)
```

#### Help

```
sessionizer --help
```

#### Example configuration

~/.zshrc

```zsh
# sessionizer
export SESSIONIZER_TARGET_LIST=~/.config/sessionizer/paths
export SESSIONIZER_LIST_GENERATOR=~/.config/sessionizer/generator
export SESSIONIZE="$HOME/projects:$HOME/work:$HOME/study"
```

~/.tmux.conf

```tmux
bind-key -r f run-shell "tmux neww -n sessionizer ~/.local/bin/sessionizer"
```

~/.config/sessionizer/generator

```sh
: > $SESSIONIZER_TARGET_LIST

for target in $(echo $SESSIONIZE | sed "s/:/ /g"); do
    find $target -mindepth 1 -maxdepth 1 -type d >> $SESSIONIZER_TARGET_LIST
done
```

## Screenshot

![screenshot](https://github.com/augustinesaidimu/sessionizer/blob/main/screenshot.png?raw=true)

## License

[MIT](https://choosealicense.com/licenses/mit/)


CHANGELOG.md
## [1.0.3](https://github.com/augustinesaidimu/sessionizer/compare/v1.0.2...v1.0.3) (2022-01-28)


### Bug Fixes

* Use a hack to fix a bug in continuation. ([1e68860](https://github.com/augustinesaidimu/sessionizer/commit/1e6886009dd78e865e1a2c7951e3566948b811ec))

## [1.0.2](https://github.com/augustinesaidimu/sessionizer/compare/v1.0.1...v1.0.2) (2022-01-28)


### Bug Fixes

* remove unused options ([295acb6](https://github.com/augustinesaidimu/sessionizer/commit/295acb6ad252189542448945eb288c18dd43178b))

## [1.0.1](https://github.com/augustinesaidimu/sessionizer/compare/v1.0.0...v1.0.1) (2022-01-28)


### Bug Fixes

* fixed bug caused by typo in variable name ([b55fa21](https://github.com/augustinesaidimu/sessionizer/commit/b55fa2198bbb74abd57b055924947f83c9c8dc35))

# 1.0.0 (2022-01-28)


### Features

* initial release ([1aeb461](https://github.com/augustinesaidimu/sessionizer/commit/1aeb461bc0f496c7a12e26f975123e80d95ec260))


install.sh
#!/usr/bin/env bash

FRAMEWORK_NAME="sessionizer"
FRAMEWORK_BRANCH="main"
FRAMEWORK_URL="https://github.com/augustinesaidimu/sessionizer"

FRAMEWORK_PATH="$HOME/.local/share/$FRAMEWORK_NAME"
FRAMEWORK_BINARY="$HOME/.local/bin/$FRAMEWORK_NAME"

_clean(){
    if [ -e "$FRAMEWORK_PATH" ]; then
        rm -rf "$FRAMEWORK_PATH"
    fi
    if [ -e "$FRAMEWORK_BINARY" ]; then
        rm -rf "$FRAMEWORK_BINARY"
    fi
}

_clone_repo() {
    sleep 0.2
    mkdir -p "$FRAMEWORK_PATH"

    git clone "$FRAMEWORK_URL" --branch="$FRAMEWORK_BRANCH" "$FRAMEWORK_PATH" --depth=1 &> /dev/null

    return $?
}

_install_script() {
    sleep 0.2
    chmod +x "$FRAMEWORK_PATH/$FRAMEWORK_NAME.sh"
    ln -sf "$FRAMEWORK_PATH/$FRAMEWORK_NAME.sh" "$FRAMEWORK_BINARY" > /dev/null  2>&1
}

_main(){
    # -- clean up --
    _clean &
    pid=$!
    _load "Initializing" $pid
    wait $pid
    [ $? -ne 0 ] && _abort

    # -- clone repo --
    _clone_repo &
    pid=$!
    _load "Fetching repo" $pid
    wait $pid
    [ $? -ne 0 ] && _abort

    # -- add command to path --
    _install_script &
    pid=$!
    _load "Installing script" $pid
    wait $pid
    [ $? -ne 0 ] && _abort

    printf "$(green "[") Installed "$FRAMEWORK_NAME" $(green "]")\n"
}

# ---------------------------------------------------------------------------- #
_abort(){
    printf "$(red "[") Install failed ! $(red "]")\n"
    exit 27
}

bold(){
  echo "\033[1;37;48m%s\033[0m" "${*}";
}

red(){
  echo "\033[1;31;48m%s\033[0m" "${*}";
}

yellow(){
  printf "\033[1;33;48m%s\033[0m" "${*}";
}

green(){
  printf "\033[1;32;48m%s\033[0m" "${*}";
}

_clear_line(){
    printf "\033[1000D" # go to begining of line
    printf "\033[0K" # clear line
}

_move_left(){
    index="$1"
    printf "\033[${index}D"
}

_move_right(){
    index="$1"
    printf "\033[${index}C"
}

_load() {
    text="$1"
    pid="$2"

    waiting=1
    while [ $waiting -eq 1 ]; do
        green "[ "
        printf "${text}    "
        green " ]"
        sleep 0.3
        _clear_line

        green "[ "
        printf "${text} .  "
        green " ]"
        sleep 0.3
        _clear_line

        green "[ "
        printf "${text} .. "
        green " ]"
        sleep 0.3
        _clear_line

        green "[ "
        printf "${text} ..."
        green " ]"

        ps cax | grep -E "\s?$pid" > /dev/null  2>&1
        if [ $? -ne 0 ]; then
            waiting=0
        fi
        sleep 0.3
        _clear_line
    done
}

# ---------------------------------------------------------------------------- #

_main $@


sessionizer.sh
#!/usr/bin/env sh
set -eu

# use default value if the are not set elsewhere
SESSIONIZER_TARGET_LIST=${SESSIONIZER_TARGET_LIST:-"${HOME}/.cache/sessionizer.targets"}
SESSIONIZER_LIST_GENERATOR=${SESSIONIZER_LIST_GENERATOR:-""}

# target list generator
_generate_list(){
  # implement your own algorithm in the SESSIONIZER_LIST_GENERATOR file

  if [ -n "$SESSIONIZER_LIST_GENERATOR" ] && [ -e "$SESSIONIZER_LIST_GENERATOR" ]; then
    # if the SESSIONIZER_LIST_GENERATOR file exists, source it
    . "$SESSIONIZER_LIST_GENERATOR"
  else
    echo "$HOME" > "$SESSIONIZER_TARGET_LIST"
  fi
}

# use fzf to list and select a session target
_select_target(){
  [ -e "$SESSIONIZER_TARGET_LIST" ] || _generate_list

  # it is possible to theme fzf
  target=$(\
    fzf --border=rounded --preview 'tree --dirsfirst -C -L 1 {}' --margin=0%\
    --color fg:#cdcecf,bg:#131A24,gutter:#131A24,hl:#f6a878,hl+:#f6a878,bg+:#283648\
    --with-nth=-2.. --delimiter="/"\
    < "$SESSIONIZER_TARGET_LIST"\
  )

  echo "$target"
}

# create a tmux session
_create_session(){
  path="$1"
  session_name="$2"

  # a script that takes the two params above. Use them as you may
  setup="$path/.project"

  # a file with enviroment variables, with the following specs:
  # entries are formatted as ENV_VARIABLE=myValue
  # comments start with a '#' on a new line.
  env="$path/.env"

  cmd="new-session -ds $session_name -c $path"

  # build the tmux options add env variables
  if [ -e "$env" ]; then
    # strip comments and blank lines from the env file
    sed -E '/^(#.*)?$/d; s/^(.*)(=)(.*)$/\1\2"\3"/g;' "$env" | while IFS= read -r var
    do
      cmd="$cmd -e $var"
    done
  fi

  # run tmux with the commands
  eval "tmux $cmd"

  # execute the setup script
  [ -x "$setup" ] && $setup "$session_name" "$path"

  # ensures continuity of script
  echo
}

# go to the selected target session
_goto_session() {
  session_name="$1"

  # if we are in a tmux session, switch to the session, otherwise attach to the
  # session

  if [ -n "$TMUX" ]; then
    tmux switch-client -t "$session_name"
  else
    tmux attach-session -t "$session_name"
  fi
}

# start a session or switch to one.
_sessionize(){
  target=$(_select_target)
  session_name=$(basename "$target" | sed -E "s/^(\.)+//; s/\./_/g" )

  # create a session if it does not exist
  if ! tmux has -t "$session_name" 2> /dev/null; then
    _create_session "$target" "$session_name"
  fi

  _goto_session "$session_name"
 }


_help(){
  cat <<EOF
$(bold RATIONALE)
  $(grey "# forget the find window.  That is for chumps")
                 from $(blue "ThePrimeagen's") .tmux.conf

$(bold USAGE)
  sessionizer [options]

$(bold ENV)
  SESSIONIZER_TARGET_LIST     a path to a file containing a list of paths.
  SESSIONIZER_LIST_GENERATOR  a shell script used to generate the list of paths.

$(bold OPTIONS)
  -h  --help     show help.
  -r  --refresh  regenerate list of session targets.

$(bold DEPENDS)
  fzf, tmux, tree, sed

$(bold SUGGESTION)
Add a binding to your tmux.conf.
  example:
  $(grey 'bind-key -r f run-shell "tmux neww ~/.local/bin/sessionizer"')

EOF
}

main(){
  option="${*}"
  case "$option" in
    -h | --help)
      _help
      ;;
    -r | --refresh)
      _generate_list
      ;;
    "" )
      _sessionize
      ;;
    * )
    {
      _help
      exit 1
    }
      ;;
  esac
}

bold(){
  printf "\033[1;37;48m%s\033[0m" "${*}"
}
blue(){
  printf "\033[0;34;48m%s\033[0m" "${*}"
}
grey(){
  printf "\033[0;37;48m%s\033[0m" "${*}"
}

main "$@"
